provider/infoblox.go

```diff
@@ -0,0 +1,384 @@
+			if _, err := strconv.Unquote(res.Text); err != nil {
+				res.Text = strconv.Quote(res.Text)
+			}
+			endpoints = append(endpoints, endpoint.NewEndpoint(res.Name, res.Text, endpoint.RecordTypeTXT))
+		}
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges applies the given changes.
+func (p *InfobloxProvider) ApplyChanges(changes *plan.Changes) error {
+	zones, err := p.zones()
+	if err != nil {
+		return err
+	}
+
+	created, deleted, updated := p.mapChanges(zones, changes)
+	p.createRecords(created)
+	p.deleteRecords(deleted)
+	p.updateRecords(updated)
+	return nil
+}
+
+func (p *InfobloxProvider) zones() ([]ibclient.ZoneAuth, error) {
+	var res, result []ibclient.ZoneAuth
+	obj := ibclient.NewZoneAuth(ibclient.ZoneAuth{})
+	err := p.client.GetObject(obj, "", &res)
+
+	if err != nil || res == nil || len(res) == 0 {
+		return nil, err
+	}
+
+	for _, zone := range res {
+		if p.domainFilter.Match(zone.Fqdn) {
+			result = append(result, zone)
+		}
+	}
+
+	return result, nil
+}
+
+type infobloxChangeMap map[*ibclient.ZoneAuth][]*endpoint.Endpoint
+
+func (p *InfobloxProvider) mapChanges(zones []ibclient.ZoneAuth, changes *plan.Changes) (infobloxChangeMap, infobloxChangeMap, infobloxChangeMap) {
+	ignored := map[string]bool{}
+	created := infobloxChangeMap{}
+	deleted := infobloxChangeMap{}
+	updated := infobloxChangeMap{}
+
+	mapChange := func(changeMap infobloxChangeMap, change *endpoint.Endpoint) {
+		zone := p.findZone(zones, change.DNSName)
+		if zone == nil {
+			if _, ok := ignored[change.DNSName]; !ok {
+				ignored[change.DNSName] = true
+				logrus.Infof("Ignoring changes to '%s' because a suitable Infoblox DNS zone was not found.", change.DNSName)
+			}
+			return
+		}
+		// Ensure the record type is suitable
+		changeMap[zone] = append(changeMap[zone], change)
+	}
+
+	for _, change := range changes.Create {
+		mapChange(created, change)
+	}
+	for _, change := range changes.Delete {
+		mapChange(deleted, change)
+	}
+	for _, change := range changes.UpdateNew {
+		mapChange(updated, change)
+	}
+	for _, change := range changes.UpdateOld {
+		mapChange(deleted, change)
+	}
+
+	return created, deleted, updated
+}
+
+func (p *InfobloxProvider) findZone(zones []ibclient.ZoneAuth, name string) *ibclient.ZoneAuth {
+	var result *ibclient.ZoneAuth
+
+	// Go through every zone looking for the longest name (i.e. most specific) as a matching suffix
+	for idx := range zones {
+		zone := &zones[idx]
+		if strings.HasSuffix(name, zone.Fqdn) {
+			if result == nil || len(zone.Fqdn) > len(result.Fqdn) {
+				result = zone
+			}
+		}
+	}
+	return result
+}
+
+func (p *InfobloxProvider) newObject(e *endpoint.Endpoint) (obj ibclient.IBObject, ref string, err error) {
+	switch e.RecordType {
+	case endpoint.RecordTypeA:
+		var records []ibclient.RecordA
+		obj = ibclient.NewRecordA(
+			ibclient.RecordA{
+				Name:     e.DNSName,
+				Ipv4Addr: e.Target,
+			},
+		)
+		p.client.GetObject(obj, "", &records)
+		if len(records) == 1 {
+			ref = records[0].Ref
```

Comment:
@mikkeloscar  I really like this idea in principle, but if `Records()` returns something other than `[]*endpoint.Endpoint` then `*provider.InfobloxProvider` won't implement `provider.Provider`. Do you have any suggestions on how to address that?

Reply:
Ofc. you are right, this I overlooked.

It's a bit stupid that you need a ref in order to delete via the infoblox API, but as a cliet you can't do much about that.

I would suggest to optimize the the `newObject` function to only make a `GetObject` in case records to be deleted. For create and update you don't need the ref so there's no need to do an extra call to the API afaict.